### ISSUE TITLE
How to keep face data (sub-material id) without interpolation using far?

### ISSUE THREAD

**Issue by FrankEscobar** [PixarAnimationStudios/OpenSubdiv#1061]
My mesh needs to keep a material id so I'll need to copy that value to the face children without interpolation.

Any tip on this?

**Comment by thomthom** [comment]
I Frank. I had similar questions like that when I implemented OSD. You can see the thread here: https://groups.google.com/forum/embed/?place=forum/opensubdiv#!topic/opensubdiv/Bre5m9zkrec

What I currently do is just traverse the faces of the final result and copy the properties from the control mesh faces. This means doing an iteration over the final faces but in my case that was not an issue.

I get the last level from the refiner, iterate all the faces for that level, then for each face get the parent face ID until I get to the control mesh face (`level->GetFaceParentFace`).

There might be some other way to copy the properties during the subdivision process, but then you'd copy the properties over to all the sub-divided faces - consuming more memory and copying more memory. 

https://groups.google.com/d/embed/msg/opensubdiv/Bre5m9zkrec/QOcAjzVQBAAJ

**Comment by barfowl** [comment]
Extracting a relevant piece of information from the thread posted above...

There is a method of the Far::PrimvarRefiner, which, despite its name, copies per-face primvar data (referred to as "uniform") rather than interpolating it, i.e. InterpolateFaceUniform().  The data types used here as arguments just need to support operator= for copying, not the Clear() and AddWithWeight() methods required for interpolation.  Other than that, its interface and usage is similar.

If this kind of bulk operation isn't appealing for some reason, you are left to inspect the hierarchy -- in which case thomthom's solution of seeking the ultimate ancestor of each refined face of interest is necessary.

At one point prior to 3.0 release we did have a refinement option to keep track of the base face index for all refined faces, but we ended up moving that functionality to PrimvarRefiner::InterpolateFaceUniform() -- which does exactly what the refinement process would do to propagate the base face indices, in addition to being able to deal directly with more general primvar data types.
